### ISSUE TITLE
Add procedure for key & cert renewal in background

### ISSUE THREAD

**Issue by gijzelaerr** [eduvpn/python-eduvpn-client#176]
a daemon or system tray icon

**Comment by maxadamo** [comment]
I'd like to see this feature, and on a side note I have a question: is there a way to expire the certificate before attempting to connect?

**Comment by gijzelaerr** [comment]
@fkooman is there an API call to expire the certificate?

We are currently working on a 2.0 release of the Linux eduvpn client which also comes with the command line interface. Using this interface it should be possible to renew a certificate on the command line, which indirectly means you can trigger this automated. Question is if this is enough for people using the GUI.

**Comment by gijzelaerr** [comment]
@maxadamo The syntax for the upcoming CLI is described in issue https://github.com/eduvpn/python-eduvpn-client/issues/229

any feedback on this prototype design is welcome.

**Comment by fkooman** [comment]
> is there a way to expire the certificate before attempting to connect?

There is no such concept. One can not "expire a certificate". A certificate _is_ expired, or not. Maybe what you mean is _revoke_. Revocation can be done through the portal on the "Account" page.

What would be the use case for revoking a certificate, especially when it is about to expire anyway?

**Comment by maxadamo** [comment]
@fkooman I should have said "revoke",I cannot shorten an expiration which was already set and I know that I can revoke it from the portal ... but it's too much time. Now, to circumvent the issue, we decided to make the certs expire in one week, otherwise, the problem was that the certificate was expiring every day, while you used the connection. 
@gijzelaerr will have a look

**Comment by gijzelaerr** [comment]
hi @maxadamo, I don't fully understand what your issue or what you are trying to do. Are you running your own server? Are you sure this is an issue with the client? What are you actually trying to do which is not possible and what change would be required to accomplish your goals?

**Comment by maxadamo** [comment]
hi @gijzelaerr 
yes, we have our own eduVPN servers and we use IDP + Tiqr to authenticate. 
1. Ideally, we'd like to authenticate with Tiqr, every time we try to connect, but it doesn't work that way . 
2. To mitigate this issue, we tried to shorten the duration of the certificate to 1 day, but after few days everyone was fed up, because the connection freezes, and we have set it to 1 week. 
3. if point 1 is not in your roadmap, it would be nice to have at least a renewal in background.

**Comment by fkooman** [comment]
> Ideally, we'd like to authenticate with Tiqr, every time we try to connect, but it doesn't work that way.

This won't be implemented.

> To mitigate this issue, we tried to shorten the duration of the certificate to 1 day, but after few days everyone was fed up, because the connection freezes on certificate expiration, and we have set expiration to 1 week.

That's why the default is 90 days. That is a more appropriate amount of time to have VPN sessions expire.

> if point 1 is not your roadmap, it would be nice to have at least a renewal in background.

I know that for the macOS and iOS apps something is being worked on at the moment to implement exactly this: 1 hour before the VPN session is about to expire the user is notified and offered to start a new authentication/authorization and reconnect with a new certificate.

**Comment by maxadamo** [comment]
> > To mitigate this issue, we tried to shorten the duration of the certificate to 1 day, but after few days everyone was fed up, because the connection freezes on certificate expiration, and we have set expiration to 1 week.
> 
> That's why the default is 90 days. That is a more appropriate amount of time to have VPN sessions expire. 

this is the tricky point for us. One could say that for 90 days in a row, your laptop password rules, and you're bypassing multi-factor authentication.

**Comment by fkooman** [comment]
> this is the tricky point for us. One could say that for 90 days in a row, your laptop password rules, and you're bypassing multi-factor authentication.

This is getting a bit off-topic now, but it really depends on your threat model. Most (web) services use 2FA only once: when logging in on a new device and then they stay logged in indefinitely. Example: GitHub. 2FA is meant as a means to be safe in case your password is compromised.

What exactly are the benefits in terms of gained security when you force users to login every day with 2FA as opposed to every 90 days?

**Comment by maxadamo** [comment]
> > this is the tricky point for us. One could say that for 90 days in a row, your laptop password rules, and you're bypassing multi-factor authentication.
> 
> This is getting a bit off-topic now, but it really depends on your threat model. Most (web) services use 2FA only once

On this aspect you're right: it's like a cookie (even the cookie of hashicorp Vault, which is a security tool, lasts forever). 

> What exactly are the benefits in terms of gained security when you force users to login every day with 2FA as opposed to every 90 days?

on one aspect, since many VPN clients offer to save your credentials, it's right what you're saying: there is no gain or little gain compared to those clients. 
But on another aspect, 2FA comes into place precisely to solve this kind of issues. 
- You can script your Google TOTP to not ask you the code and store the secret on your laptop
- you can forget your yubikey inserted in your laptop
- but you are never gonna loose your phone with Tiqr and you're not gonna loose your finger along with your laptop and your phone.  

_Hence the question_: why trying to gain so much security with Tiqr and finger-print, if everyone can gain access all the times?

**Comment by fkooman** [comment]
I think pretending 2FA protects against anything more than a compromised password is a bad idea.

Also, it is up to the individual server operators to configure expiry times. Nobody tells them to set it to 1 day or 1 week. The default of 90 days seems reasonable as that is usually the amount of time students still have access to campus services after they graduate.

In the admin portal one can also always manually disable the user with immediate effect after a compromise or when they no longer are part of the organization.

So I think the questions you ask are not so much meant for me, but more for your server operator to answer. 

I am _still_ not clear on what the exact threat is you are trying to protect against with 2FA every day, except compromised password...

**Comment by maxadamo** [comment]
> Also, it is up to the individual server operators to configure expiry times. Nobody tells them to set it to 1 day or 1 week. The default of 90 days seems reasonable as that is usually the amount of time students still have access to campus services after they graduate. 

that's one point (which I was gonna mention in my previous messages): we use to work with academic institutions, but in this case we are using eduVPN for corporate access. That's why our security department is a bit pushy with security

> So I think the questions you ask are not so much meant for me, but more for your server operator to answer.
> 
> I am _still_ not clear on what the exact threat is you are trying to protect against with 2FA every day, except compromised password...

I'm the DevOps. I configured our VPN servers, I handed them over to our IT, and I let the security and IT departments take the decisions on these settings and I can agree with you on some aspects, but I mostly report the opinions of our departments... I struggled with them to raise the expiration from 1 day to 1 week.

**Comment by fkooman** [comment]
> That's why our security department is a bit pushy with security.

Yes. But what does this mean _concretely_? What security? Whose security? What are you protecting against? How does reducing `sessionExpiry` help with this? I'm really curious now!

**Comment by gijzelaerr** [comment]
a CLI `refresh` subcommand has been implemented in the 2.0 branch. Normally I should close this issue, but it seems there is still an active discussion going on. Maybe an idea to move this to another issue tracker?

thanks!

**Comment by fkooman** [comment]
what does `refresh` do exactly?

**Comment by fkooman** [comment]
> what does `refresh` do exactly?

```
$ eduvpn refresh

 - Refreshes active configuration, refreshes tokens and updates certificate if expires
 
 - non-interactive, but blocks when browser action is required. Not sure if this is a good idea if started from cron
```

FYI: updating (`/create_keypair`) certificates only? makes sense after reauthorization, otherwise the cert will expire at exactly the same moment...

**Comment by maxadamo** [comment]
I entirely disagree the term of the discussion: I asked if a feature was in the road-map. 
It isn't. End of story for me. What is with all this noise? 
You argue because my organization would like to re-auth on every connection attempt, which is the most common setup for a VPN (unless it's a site2site VPN and the certificate are properly secured, and it's not a laptop that you leave somewhere open), and it's the logical consequence of he adoption of 2FA. But I am not pushing anything. I was only wondering. 
I have tried to explain that someone in my organization had decided to shorten the validity of the certificate, which doesn't mean that I agree this decision. 
You don't even understand what people are trying to explain, and I encourage you to improve your attitude.

**Comment by maxadamo** [comment]
having said that, on the user experience aspect, this idea looks good to me.

**Comment by fkooman** [comment]
Actually, what I was trying to do was ask you the questions you can use to talk to your security people. Which would result in either:

1. Reconsideration of the short expiry because they agree that it doesn't really add anything to improve their vision on security, making your existing problem less urgent;
2. They come up with a threat model (scenario) that we didn't think of, which could result in us resolving the threat, or recommending shorter lived certificates for *all* servers.

The point of this "noise" is, at least for me, to try and learn something new and possibly improving the service for everyone.

**Comment by gijzelaerr** [comment]
i think we can close this issue now.
